src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs

```diff
@@ -121,4 +127,55 @@ public abstract partial class DiagnosticSource
         public System.Diagnostics.Activity StartActivity(System.Diagnostics.Activity activity, object? args) { throw null; }
         public void StopActivity(System.Diagnostics.Activity activity, object? args) { }
     }
+    public enum ActivitySourceEventOperation
+    {
+        SourceCreated = 0,
+        ActivityStarted = 1,
+        ActivityStopped = 2
+    }
+    public enum ActivityKind
+    {
+        Internal = 1,
+        Server = 2,
+        Client = 3,
+        Producer = 4,
+        Consumer = 5,
+    }
+    public sealed class ActivitySourceEventArgs : System.EventArgs
+    {
+        private ActivitySourceEventArgs() { throw null; }
+        public System.Diagnostics.ActivitySourceEventOperation Operation { get; }
+    }
+    public readonly struct ActivityContext : IEquatable<ActivityContext>
+    {
+        public ActivityContext(System.Diagnostics.ActivityTraceId traceId, System.Diagnostics.ActivitySpanId spanId, System.Diagnostics.ActivityTraceFlags traceOptions, string? traceState = null) { throw null; }
+        public System.Diagnostics.ActivityTraceId TraceId { get; }
+        public System.Diagnostics.ActivitySpanId SpanId { get; }
+        public System.Diagnostics.ActivityTraceFlags TraceFlags { get; }
+        public string? TraceState { get; }
+        public static bool operator ==(System.Diagnostics.ActivityContext context1, System.Diagnostics.ActivityContext context2) { throw null; }
+        public static bool operator !=(System.Diagnostics.ActivityContext context1, System.Diagnostics.ActivityContext context2) { throw null; }
+        public bool Equals(System.Diagnostics.ActivityContext context) { throw null; }
+        public override bool Equals(object? obj) { throw null; }
+        public override int GetHashCode() { throw null; }
+    }
+    public readonly struct ActivityLink
+    {
+        public ActivityLink(System.Diagnostics.ActivityContext context) { throw null; }
+        public ActivityLink(System.Diagnostics.ActivityContext context, System.Collections.Generic.IDictionary<string, object>? attributes) { throw null; }
+        public System.Diagnostics.ActivityContext Context { get; }
+        public System.Collections.Generic.IDictionary<string, object>? Attributes { get; }
+    }
+    public sealed class ActivitySource : IDisposable
+    {
+        private ActivitySource() { throw null; }
+        public ActivitySource(string name) { throw null; }
+        public static event EventHandler<System.Diagnostics.ActivitySourceEventArgs>? OperationEvent { add { } remove { } }
```

Comment:
I've been investigating this a bit more and had some thoughts:

1) The IObservable<ActivitySource> pattern doesn't work well here so I recommend we eliminate it from consideration. It forces the observer to cache every ActivitySource it subscribes to in order to unsubscribe later. That list wastes a bunch of memory and probably prevents any ActivitySource from ever getting collected. Technically the subscriber could use a second IObserver to iterate the list a second time to unsubscribe but it makes the code quite complex and I don't believe there is a way to be certain the unsubscribe pass is complete without making assumptions about threading mechanics and the BCL's implementation.

2) If we are moving to ActivityListener as an abstract base class then there is another option which seems appealing:

````
abstract class ActivityListener : IDisposable
{
    // Begins the flow of callback notifications
    public void Subscribe();

    // Unsubscribes from all activity notifications
    public void Dispose() { ... }

    // Derived implementations can override this to filter which activity names they will listen to.
    // Only names where this function returns true will ever trigger Sampling/Start/Stop callbacks
    // Internally the BCL would invoke this callback for all existing sources inside of Subscribe() and
    // again in the future whenever new sources are created. If the function returns true then the BCL
    // adds this listener to the source.
    public virtual bool ShouldListenToActivity(string name) { return true; }

    // activity callbacks
    public virtual bool ShouldCreateActivity(string name, ActivityTraceId id, ...) { return true; }
    public virtual void OnActivityStarted(Activity a) {}
    public virtual void OnActivityStopped(Activity a) {}
}
````
To use it the consuming code derives from the ActivityListener, instantiates an instance of the derived type and then calls Subscribe() to begin the flow of callbacks. This is very similar to the EventSource/EventListener pattern except I added an explicit Subscribe() step to ensure that events don't start showing up before the derived constructor has had a chance to run.

If we went this route we could eliminate these APIs because subscription is now automatic:
static ActivitySource.ActiveList
static ActivitySource.OperationEvent
ActivitySource.ActivityEvent

Reply:
@noahfalk I have applied your suggestion. The only small change I did is I have moved the Subscribe method from the listener to ActivitySource and named it AddListener for the sake of discoverability. Let me know if you have any more feedback.